### ISSUE TITLE
[Web] fp16 and q4f16 Gemma 3 models produce invalid outputs on WebGPU due to overflow in ONNX runtime

### ISSUE THREAD

**Issue by fosple** [microsoft/onnxruntime#26732]
### Describe the issue

When running Gemma 3 (270M) models exported to ONNX produces unusable outputs using WebGPU and fp16.

The same model work correctly on:

* ONNX Runtime CPU / WASM
* WebGPU when using fp32

This strongly suggests an overflow / numerical stability issue inside ONNX Runtime’s WebGPU fp16 execution path, likely identical or related to issue [[WebGPU] fp16 nanochat produces NaNs (CPU works fine)](https://github.com/microsoft/onnxruntime/issues/26367)

Maybe this has also to do with the [unsloth finding](https://docs.unsloth.ai/models/gemma-3-how-to-run-and-fine-tune#gemma-3-fixes-analysis) that activations become infinity for `float16`?

### To reproduce

The failure can be reproduced with a minimal Transformers.js script:

`main.js`:
```
import { pipeline } from "@huggingface/transformers";

const generator = await pipeline(
  "text-generation",
  "onnx-community/gemma-3-270m-it-ONNX",
  { 
    dtype: 'fp16',
    device: 'webgpu',
  },
);

const messages = [
  { role: "system", content: "You are a helpful assistant." },
  { role: "user", content: "What is the capital of France?" },
];

const output = await generator(messages, { max_new_tokens: 128 });
console.log(output[0].generated_text.at(-1).content);
```

HTML to try it out:
```
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Transformers.js Browser Demo</title>
  <script type="importmap">
  {
    "imports": {
      "@huggingface/transformers": "https://cdn.jsdelivr.net/npm/@huggingface/transformers"
    }
  }
  </script>
</head>
<body>
  <h1>Transformers.js Browser Demo</h1>
  <p>Open the browser console to see the generated output.</p>
  <script type="module" src="./main.js"></script>
</body>
</html>
```

For model `HuggingFaceTB/SmolLM2-360M-Instruct` with `dtype: 'fp16'` and `device: 'webgpu'` it returns:
```
The capital of France is Paris.
```

For model `onnx-community/gemma-3-270m-it-ONNX` with `dtype: 'fp16'` and `device: 'webgpu'` it returns:
```
<unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56>
```

For model `onnx-community/gemma-3-270m-it-ONNX` with `dtype: 'fp16'` and `device: 'wasm'` it returns:
```
The capital of France is Paris.
```

For model  `onnx-community/gemma-3-270m-it-ONNX` with `dtype: 'q4f16'` and `device: 'webgpu'` it returns:
```
<unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56><unused56>
```

For model  `onnx-community/gemma-3-270m-it-ONNX` with `dtype: 'q4f16'` and `device: 'wasm'` it returns:
```
The capital of France is Paris.
```

### Urgency

High.

Gemma-3-270M is designed for edge and client-side inference, especially WebGPU.
Because FP16 and Q4F16 are the intended fast modes, this overflow prevents the model from being usable on ONNX Runtime WebGPU in real deployments.

This affects:

* ONNX Runtime-based inference pipelines in browsers
* Transformers.js WebGPU backend
* Any WebGPU client expecting FP16 performance

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.23.2

### Execution Provider

'webgpu' (WebGPU)

**Comment by xenova** [comment]
Hi 👋 I recently updated the float16 versions at https://huggingface.co/onnx-community/gemma-3-270m-it-ONNX/commit/cfd5c04f84a64766d63efc5bb1d2cf31f34a4a90; could you clear model cache (via devtools) and try again?

**Comment by fosple** [comment]
@xenova Nice 🎉 Just tested it extensively and it seams to work really well. It's also really fast on WebGPU. Thanks so much for having a look into it ❤️

## Speed / RAM

- fp32 & webgpu - 27.78 token/sec @ 4 GB RAM
- fp16 & webgpu -> 31.30 token/sec @ 3 GB RAM
- q4f16 & webgpu -> 38.34 tok/sec @ 2.5 GB RAM

I ran speed test on Chrome / MacBook with M2 processor.

Pretty awesome speed 🚀

## Quality

Output text quality is mixed, but can't really tell without having fine-tuned the model. 

I was wondering if you used <https://huggingface.co/google/gemma-3-270m-it> or <https://huggingface.co/google/gemma-3-270m-it-qat-q4_0-unquantized> for the conversion to onnx? `gemma-3-270m-it-qat-q4_0-unquantized` might be a better fit for the quant onnx files :)

**Comment by xenova** [comment]
Glad to hear it's working well :)

> I was wondering if you used https://huggingface.co/google/gemma-3-270m-it or https://huggingface.co/google/gemma-3-270m-it-qat-q4_0-unquantized for the conversion to onnx? gemma-3-270m-it-qat-q4_0-unquantized might be a better fit for the quant onnx files :)

Indeed, I used the non-QAT checkpoints for the conversion, and I haven't tested on the QAT versions 😅 very possible that it could improve performance. I'm aware of some research from the microsoft team about this too (https://arxiv.org/abs/2501.06417), which may be useful to look into... if I ever get time haha.

**Comment by fosple** [comment]
@xenova Would be awesome if you could share your conversion script? I would really like to try it out with my fine tuned models - as gemma 270m is mostly only usable after fine tuning due to it's limited general knowledge 🙂

**Comment by xenova** [comment]
@fosple Sure! We collaborated with google to release a conversion script for the model. You can find the colab notebook at https://colab.research.google.com/github/google-gemini/gemma-cookbook/blob/main/Demos/Emoji-Gemma-on-Web/resources/Convert_Gemma_3_270M_to_ONNX.ipynb#scrollTo=Cn2YgWGrNv2M

**Comment by rgryta** [comment]
Noticed that this likely is the cause of kokoro-js not working with webgpu backend too 😦

**Comment by RobertoReale** [comment]
## Root cause: the fp16 **residual stream** overflows, not the matmul kernels

This issue was auto-closed by the stale bot, and separately I had misattributed it — in #29599 I claimed it was an f16 **accumulator** overflow inside the WebGPU MatMul/MatMulNBits kernels. I've now measured it properly on the model, and that was wrong. Posting the correct diagnosis here so it's on the record, because @fosple's own hunch in the original report turns out to be the right one.

### Measurement

`onnx-community/gemma-3-270m-it-ONNX` pinned to `2950c41f` — the last revision *before* the 2025-12-02 re-export, i.e. the one this issue was filed against. Real prompt, real activations captured on the CPU EP, real q4 weights dequantized from the model file.

**1. The MatMulNBits accumulators never overflow.** Accumulating in f16 in the kernel's order (block32 partial sums carried across K), across **all 126 MatMulNBits nodes**, the peak running partial sum is **~324** — three orders of magnitude below the f16 ceiling of 65504. `down_proj`, the only K=2048 matmul in the model and the one whose input (`gelu(gate) * up`) is not normalized, peaks at **37**.

**2. The residual stream does overflow.** Traced in fp16, it grows monotonically down the stack and crosses 65504 between layers 7 and 8:

| tensor | max abs (fp16) |
|---|---:|
| `L5/Add_2` | 40,000 |
| `L6/Add_2` | 52,096 |
| `L7/Add_1` | 52,352 |
| `L7/Add_2` | **Inf** |
| `L8/Add_1` | 65,472 |

Once it saturates, NaN propagates through the following RMSNorm/Softmax and the output degenerates — the empty/garbage generation reported here.

**3. The Hub-side mitigation confirms it.** The 2025-12-02 re-export (commit `cfd5c04f84`, *"update fp16 models with clips"*) added **18 `Clip` nodes — one per layer, on the residual adds `/model/layers.N/Add_1`, bounded at ±32752** (exactly half of 65504, leaving headroom for the next residual add). Not one of them is on a matmul output. That is a targeted fix for a residual-range overflow, and it is why the current Hub model no longer reproduces.

This is the known Gemma-3 fp16 activation-range behaviour — exactly the [unsloth finding](https://docs.unsloth.ai/models/gemma-3-how-to-run-and-fine-tune#gemma-3-fixes-analysis) @fosple linked in the original report.

### Why CPU/WASM and WebGPU-fp32 are unaffected

Both carry the residual in f32 (MLAS accumulates and stores intermediates in float even for f16 inputs; the fp32 model obviously has the range). The WebGPU fp16 path keeps the residual in f16 end-to-end, so a value that genuinely exceeds 65504 actually saturates there. The bug is a **dynamic-range** problem in the model, surfaced only by the backend that honours fp16 storage.

### What this means

- **#29599 (f32 matmul accumulators) does not fix this issue.** I've removed the `Fixes #26732` claim from it. It addresses a real but *different* overflow (demonstrated end-to-end on whisper-small q4), and it would not have saved Gemma 3.
- **The model-side fix already shipped**, so users pulling the current Hub revision are fine. Anyone pinning an older revision, or exporting Gemma 3 to fp16 themselves, will hit this again.
- **If ORT wants a runtime-side answer**, the lever is keeping the residual/skip path in f32 on the WebGPU EP for fp16 models (the `SkipSimplifiedLayerNormalization` fusion is the natural place — it already reads the residual, so promoting that carry to f32 costs no extra bandwidth on the weights), rather than requiring every fp16 model with large activations to ship Clip nodes. That's a design call for the WebGPU EP owners — cc @guschmue @fs-eire @qjia7 — and I'm happy to prototype it if there's interest.

Repro scripts (load the ONNX, expose intermediates, run the CPU EP, dequantize, simulate the kernel's accumulation order) are straightforward to reproduce from the numbers above; happy to attach them if useful.
